File: resources/ios/Sources/MobikulAppBarFunctions.swift
import Foundation
import NativePHP

enum MobikulAppBarFunctions {
    private static let allowedIcons: Set<String> = ["back", "menu", "search", "cart", "profile", "wishlist", "share", "filter", "notifications", "more"]

    struct Configure: BridgeFunction {
        func run(parameters: [String: Any], resolve: ResolveBlock, reject: RejectBlock) {
            let title = (parameters["title"] as? String)?.trimmingCharacters(in: .whitespacesAndNewlines)
            let context = (parameters["context"] as? String)?.trimmingCharacters(in: .whitespacesAndNewlines)

            let finalTitle = (title?.isEmpty == false ? title! : "App")
            let finalContext = (context?.isEmpty == false ? context! : "default")
            let icons = MobikulAppBarFunctions.normalizeIcons(parameters["icons"], context: finalContext)
            let options = MobikulAppBarFunctions.normalizeOptions(parameters["options"])

            resolve([
                "title": finalTitle,
                "context": finalContext,
                "icons": icons,
                "options": options,
                "updated": true
            ])
        }
    }

    struct Reset: BridgeFunction {
        func run(parameters: [String: Any], resolve: ResolveBlock, reject: RejectBlock) {
            resolve([
                "title": "App",
                "context": "default",
                "icons": MobikulAppBarFunctions.defaultIcons(for: "default"),
                "options": MobikulAppBarFunctions.defaultOptions(),
                "updated": true
            ])
        }
    }

    private static func normalizeIcons(_ rawIcons: Any?, context: String) -> [[String: String]] {
        guard let rawList = rawIcons as? [[String: Any]] else {
            return defaultIcons(for: context)
        }

        let filtered = rawList.compactMap { item -> [String: String]? in
            guard let icon = (item["icon"] as? String)?.trimmingCharacters(in: .whitespacesAndNewlines),
                  let action = (item["action"] as? String)?.trimmingCharacters(in: .whitespacesAndNewlines),
                  !icon.isEmpty,
                  !action.isEmpty,
                  allowedIcons.contains(icon) else {
                return nil
            }

            return ["icon": icon, "action": action]
        }

        return filtered.isEmpty ? defaultIcons(for: context) : filtered
    }

    private static func defaultIcons(for context: String) -> [[String: String]] {
        switch context.lowercased() {
        case "home":
            return [["icon": "menu", "action": "open_menu"], ["icon": "search", "action": "open_search"], ["icon": "cart", "action": "open_cart"]]
        case "product":
            return [["icon": "back", "action": "go_back"], ["icon": "wishlist", "action": "add_to_wishlist"], ["icon": "share", "action": "share_product"]]
        case "checkout":
            return [["icon": "back", "action": "go_back"], ["icon": "cart", "action": "open_cart"]]
        case "profile":
            return [["icon": "back", "action": "go_back"], ["icon": "notifications", "action": "open_notifications"], ["icon": "more", "action": "open_more"]]
        default:
            return [["icon": "menu", "action": "open_menu"], ["icon": "search", "action": "open_search"]]
        }
    }

    private static func normalizeOptions(_ rawOptions: Any?) -> [String: Any] {
        let options = rawOptions as? [String: Any] ?? [:]
        let defaults = defaultOptions()

        return [
            "showAppLogo": booleanOption(options, key: "showAppLogo", fallback: defaults["showAppLogo"] as? Bool ?? false),
            "isElevated": booleanOption(options, key: "isElevated", fallback: defaults["isElevated"] as? Bool ?? true),
            "isLeadingEnable": booleanOption(options, key: "isLeadingEnable", fallback: defaults["isLeadingEnable"] as? Bool ?? false),
            "isHomeEnable": booleanOption(options, key: "isHomeEnable", fallback: defaults["isHomeEnable"] as? Bool ?? false),
            "isAppLogoForDarkmode": booleanOption(options, key: "isAppLogoForDarkmode", fallback: defaults["isAppLogoForDarkmode"] as? Bool ?? false),
            "appLogoUrl": stringOption(options, key: "appLogoUrl") ?? "",
            "darkAppLogoUrl": stringOption(options, key: "darkAppLogoUrl") ?? "",
            "placeHolderImage": stringOption(options, key: "placeHolderImage") ?? "",
            "appBarBackgroundColor": colorOption(options, key: "appBarBackgroundColor", fallback: defaults["appBarBackgroundColor"] as? String ?? "#ffffff"),
            "titleColor": colorOption(options, key: "titleColor", fallback: defaults["titleColor"] as? String ?? "#111827"),
            "titleFontSize": intOption(options, key: "titleFontSize", fallback: defaults["titleFontSize"] as? Int ?? 18, minimum: 12),
            "titleFontWeight": stringOption(options, key: "titleFontWeight") ?? "600",
            "logoWidth": intOption(options, key: "logoWidth", fallback: defaults["logoWidth"] as? Int ?? 32, minimum: 20),
            "logoHeight": intOption(options, key: "logoHeight", fallback: defaults["logoHeight"] as? Int ?? 32, minimum: 20)
        ]
    }

    private static func defaultOptions() -> [String: Any] {
        return [
            "showAppLogo": false,
            "isElevated": true,
            "isLeadingEnable": false,
            "isHomeEnable": false,
            "isAppLogoForDarkmode": false,
            "appLogoUrl": "",
            "darkAppLogoUrl": "",
            "placeHolderImage": "",
            "appBarBackgroundColor": "#ffffff",
            "titleColor": "#111827",
            "titleFontSize": 18,
            "titleFontWeight": "600",
            "logoWidth": 32,
            "logoHeight": 32
        ]
    }

    private static func booleanOption(_ options: [String: Any], key: String, fallback: Bool) -> Bool {
        return options[key] as? Bool ?? fallback
    }

    private static func stringOption(_ options: [String: Any], key: String) -> String? {
        guard let value = options[key] as? String else {
            return nil
        }

        let trimmed = value.trimmingCharacters(in: .whitespacesAndNewlines)
        return trimmed.isEmpty ? nil : trimmed
    }

    private static func intOption(_ options: [String: Any], key: String, fallback: Int, minimum: Int) -> Int {
        let value: Int

        switch options[key] {
        case let number as Int:
            value = number
        case let number as Double:
            value = Int(number)
        default:
            value = fallback
        }

        return max(minimum, value)
    }

    private static func colorOption(_ options: [String: Any], key: String, fallback: String) -> String {
        guard let value = stringOption(options, key: key) else {
            return fallback
        }

        let hexPattern = "^#[0-9a-fA-F]{3,8}$"
        let cssFunctionPattern = "^(rgb|rgba|hsl|hsla)\\([^)]+\\)$"

        if value.range(of: hexPattern, options: .regularExpression) != nil {
            return value
        }

        if value.range(of: cssFunctionPattern, options: .regularExpression) != nil {
            return value
        }

        return fallback
    }
}

File: resources/android/src/MobikulAppBarFunctions.kt
package com.mobikul.plugins.appbar

import androidx.fragment.app.FragmentActivity
import com.nativephp.mobile.bridge.BridgeFunction
import com.nativephp.mobile.bridge.BridgeResponse

object MobikulAppBarFunctions {
    private val allowedIcons = setOf("back", "menu", "search", "cart", "profile", "wishlist", "share", "filter", "notifications", "more")

    class Configure(private val activity: FragmentActivity) : BridgeFunction {
        override fun execute(parameters: Map<String, Any>): Map<String, Any> {
            val title = (parameters["title"] as? String)?.trim().orEmpty().ifEmpty { "App" }
            val context = (parameters["context"] as? String)?.trim().orEmpty().ifEmpty { "default" }
            val icons = normalizeIcons(parameters["icons"], context)
            val options = normalizeOptions(parameters["options"])

            return BridgeResponse.success(
                mapOf(
                    "title" to title,
                    "context" to context,
                    "icons" to icons,
                    "options" to options,
                    "updated" to true
                )
            )
        }
    }

    class Reset(private val activity: FragmentActivity) : BridgeFunction {
        override fun execute(parameters: Map<String, Any>): Map<String, Any> {
            return BridgeResponse.success(
                mapOf(
                    "title" to "App",
                    "context" to "default",
                    "icons" to defaultIcons("default"),
                    "options" to defaultOptions(),
                    "updated" to true
                )
            )
        }
    }

    private fun normalizeIcons(rawIcons: Any?, context: String): List<Map<String, String>> {
        val normalized = mutableListOf<Map<String, String>>()

        if (rawIcons is List<*>) {
            rawIcons.forEach { item ->
                val map = item as? Map<*, *> ?: return@forEach
                val icon = (map["icon"] as? String)?.trim().orEmpty()
                val action = (map["action"] as? String)?.trim().orEmpty()

                if (icon.isNotEmpty() && action.isNotEmpty() && allowedIcons.contains(icon)) {
                    normalized.add(mapOf("icon" to icon, "action" to action))
                }
            }
        }

        return if (normalized.isEmpty()) defaultIcons(context) else normalized
    }

    private fun defaultIcons(context: String): List<Map<String, String>> {
        return when (context.lowercase()) {
            "home" -> listOf(mapOf("icon" to "menu", "action" to "open_menu"), mapOf("icon" to "search", "action" to "open_search"), mapOf("icon" to "cart", "action" to "open_cart"))
            "product" -> listOf(mapOf("icon" to "back", "action" to "go_back"), mapOf("icon" to "wishlist", "action" to "add_to_wishlist"), mapOf("icon" to "share", "action" to "share_product"))
            "checkout" -> listOf(mapOf("icon" to "back", "action" to "go_back"), mapOf("icon" to "cart", "action" to "open_cart"))
            "profile" -> listOf(mapOf("icon" to "back", "action" to "go_back"), mapOf("icon" to "notifications", "action" to "open_notifications"), mapOf("icon" to "more", "action" to "open_more"))
            else -> listOf(mapOf("icon" to "menu", "action" to "open_menu"), mapOf("icon" to "search", "action" to "open_search"))
        }
    }

    private fun normalizeOptions(rawOptions: Any?): Map<String, Any?> {
        val options = rawOptions as? Map<*, *> ?: emptyMap<Any, Any>()
        val defaults = defaultOptions()

        return mapOf(
            "showAppLogo" to booleanOption(options, "showAppLogo", defaults["showAppLogo"] as Boolean),
            "isElevated" to booleanOption(options, "isElevated", defaults["isElevated"] as Boolean),
            "isLeadingEnable" to booleanOption(options, "isLeadingEnable", defaults["isLeadingEnable"] as Boolean),
            "isHomeEnable" to booleanOption(options, "isHomeEnable", defaults["isHomeEnable"] as Boolean),
            "isAppLogoForDarkmode" to booleanOption(options, "isAppLogoForDarkmode", defaults["isAppLogoForDarkmode"] as Boolean),
            "appLogoUrl" to stringOption(options, "appLogoUrl"),
            "darkAppLogoUrl" to stringOption(options, "darkAppLogoUrl"),
            "placeHolderImage" to stringOption(options, "placeHolderImage"),
            "appBarBackgroundColor" to colorOption(options, "appBarBackgroundColor", defaults["appBarBackgroundColor"] as String),
            "titleColor" to colorOption(options, "titleColor", defaults["titleColor"] as String),
            "titleFontSize" to intOption(options, "titleFontSize", defaults["titleFontSize"] as Int, 12),
            "titleFontWeight" to (stringOption(options, "titleFontWeight") ?: defaults["titleFontWeight"]),
            "logoWidth" to intOption(options, "logoWidth", defaults["logoWidth"] as Int, 20),
            "logoHeight" to intOption(options, "logoHeight", defaults["logoHeight"] as Int, 20)
        )
    }

    private fun defaultOptions(): Map<String, Any> {
        return mapOf(
            "showAppLogo" to false,
            "isElevated" to true,
            "isLeadingEnable" to false,
            "isHomeEnable" to false,
            "isAppLogoForDarkmode" to false,
            "appLogoUrl" to "",
            "darkAppLogoUrl" to "",
            "placeHolderImage" to "",
            "appBarBackgroundColor" to "#ffffff",
            "titleColor" to "#111827",
            "titleFontSize" to 18,
            "titleFontWeight" to "600",
            "logoWidth" to 32,
            "logoHeight" to 32
        )
    }

    private fun booleanOption(options: Map<*, *>, key: String, fallback: Boolean): Boolean {
        return options[key] as? Boolean ?: fallback
    }

    private fun stringOption(options: Map<*, *>, key: String): String? {
        val value = (options[key] as? String)?.trim().orEmpty()
        return if (value.isEmpty()) null else value
    }

    private fun intOption(options: Map<*, *>, key: String, fallback: Int, minimum: Int): Int {
        val value = when (val raw = options[key]) {
            is Int -> raw
            is Double -> raw.toInt()
            is Float -> raw.toInt()
            else -> fallback
        }

        return maxOf(minimum, value)
    }

    private fun colorOption(options: Map<*, *>, key: String, fallback: String): String {
        val value = stringOption(options, key) ?: return fallback
        val hexRegex = Regex("^#[0-9a-fA-F]{3,8}$")
        val cssFunctionRegex = Regex("^(rgb|rgba|hsl|hsla)\\([^)]+\\)$")

        return if (hexRegex.matches(value) || cssFunctionRegex.matches(value)) value else fallback
    }
}
